crates/shared/src/price_estimation/oneinch.rs

```diff
@@ -0,0 +1,236 @@
+use super::gas;
+use crate::oneinch_api::{OneInchClient, RestResponse, SellOrderQuoteQuery};
+use crate::price_estimation::{Estimate, PriceEstimating, PriceEstimationError, Query};
+use futures::future;
+use model::order::OrderKind;
+use primitive_types::U256;
+use std::sync::Arc;
+
+pub struct OneInchPriceEstimator {
+    pub api: Arc<dyn OneInchClient>,
+}
+
+impl OneInchPriceEstimator {
+    async fn estimate(&self, query: &Query) -> Result<Estimate, PriceEstimationError> {
+        if query.kind == OrderKind::Buy {
+            return Err(PriceEstimationError::UnsupportedOrderType);
+        }
+
+        let quote = self
+            .api
+            .get_sell_order_quote(SellOrderQuoteQuery {
+                from_token_address: query.sell_token,
+                to_token_address: query.buy_token,
+                amount: query.in_amount,
+                protocols: None,
+                fee: None,
+                gas_limit: None,
+                connector_tokens: None,
+                complexity_level: None,
+                main_route_parts: None,
+                virtual_parts: None,
+                parts: None,
+                gas_price: None,
+            })
+            .await
+            .map_err(PriceEstimationError::Other)?;
+
+        match quote {
+            RestResponse::Ok(quote) => Ok(Estimate {
+                out_amount: quote.to_token_amount,
+                gas: U256::from(gas::SETTLEMENT_OVERHEAD) + quote.estimated_gas,
+            }),
+            RestResponse::Err(e) => {
+                Err(PriceEstimationError::Other(anyhow::anyhow!(e.description)))
+            }
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl PriceEstimating for OneInchPriceEstimator {
+    async fn estimates(
+        &self,
+        queries: &[Query],
+    ) -> Vec<anyhow::Result<Estimate, PriceEstimationError>> {
+        debug_assert!(queries.iter().all(|query| {
```

Comment:
Maybe add a message to the `debug_assert` to add some context as to why this should never happen? (The way I see it this is because it should always be wrapped in `SanitizedPriceEstimator` which guarantees these invariants, but it may not be obvious to someone less familiar with this code).